### ISSUE TITLE
Add support for NVIDIA Optical Flow Accelerator (NVOFA) FRUC

### ISSUE THREAD

**Issue by OpenSourceAnarchist** [mpv-player/mpv#11060]
### Expected behavior of the wanted feature
MPV already ships with CUDA support I believe. All this would require, then, is adding in libraries for NVIDIA Optical Flow SDK 4.0, which includes the NVOFA assisted Frame Rate Up Conversion (FRUC) library. The library exposes FRUC APIs that can be used for frame rate up conversion of video files. See [https://docs.nvidia.com/video-technologies/optical-flow-sdk/nvfruc-programming-guide/index.html](https://docs.nvidia.com/video-technologies/optical-flow-sdk/nvfruc-programming-guide/index.html) for the API documentation and sample code.

This is important because NVIDIA GPUs since the Turing generation have hardware support for optical flow-based frame interpolation. MPV already includes `--interpolation`, but if the GPU supports optical flow-based frame interpolation, this should obviously be preferred for maximum smoothness and quality. I believe NVOFA FRUC only returns interpolated frames, so exactly which frames are interpolated would need to be decided (to match the user's monitor's frame rate, to match to an arbitrary fps, to just interpolate between all frames --> 2x fps, etc.).
### Alternative behavior of the wanted feature
Even a "dumb" implementation that simply interpolates all frames (2x total frames) would be highly appreciated. Effectively, this would eliminate the need to use [SVP](https://www.svp-team.com/) since MPV itself would be capable of hardware-based frame interpolation for NVIDIA GPU owners!

**Comment by philipl** [comment]
This is something I've been thinking about doing, but two things:
- It would be an FFmpeg filter, and not actually any code in mpv
- It would almost certainly count as `nonfree` in FFmpeg and so be unredistributable in binary form if it was even considered acceptable to merge.

The api itself says it can only accept nv12 and rgb, which would make proper handling of 10/12 bit content impossible, as you really don't want to do interpolation on the final scaled rgb and neither do you want to do rgb conversion in a filter as opposed to in the `vo`.

**Comment by OpenSourceAnarchist** [comment]
Thanks for clarifying those points. Should I make a bug report on FFmpeg's tracker and link it here since this issue can't be fulfilled  until the filter gets upstreamed? I can mention the caveat with 10/12 bit video files.

I guess the only code in mpv needed would then be a build flag to include support for NVOFA FRUC and a runtime flag to enable/disable it (and perhaps what GPU device you want it to run on). Or since it's just a FFmpeg filter, it could be left to the user to manually specify the filter and any parameters which would be passed along to FFmpeg.

**Comment by philipl** [comment]
There's no point creating an ffmpeg bug. I know that I'm trying this out and no one else will pay any attention to it.

Anyway, the whole thing is horribly written with poor documentation. It currently segfaults in an undebuggable way and I'm seeing if I can get any help from nvidia to understand why it's not working.

**Comment by philipl** [comment]
Ok. I've got a working [implementation](https://github.com/philipl/FFmpeg/tree/fruc-me) based on `vf_framerate` (which is a simpler blending filter). As noted, the licensing here is dubious - I do not believe any resulting binaries are distributable, and you need to provide libNvOFFRUC.so and libcuda-rt in your library path. But feel free to try it out.

```
mpv --vf=nvoffruc:fps=120 somefile.mkv
```

**Comment by philipl** [comment]
I've cleaned up the change on my branch. Feel free to try it out. It's never going to be mergeable given the licensing, but you can build it and use it locally if you want. I'm going to close this issue as that's the closest you're going to get under the current circumstances.

**Comment by Usulyre** [comment]
> I've cleaned up the change on my branch. Feel free to try it out. It's never going to be mergeable given the licensing, but you can build it and use it locally if you want. I'm going to close this issue as that's the closest you're going to get under the current circumstances.

Hi a bit off topic here, but do you have a ffmpeg binary to try out the filter with? Also, how does one specify the parameters in ffmpeg for the filter?

**Comment by Jules-A** [comment]
AMD released the FSR3 source code, I wonder if it would be more practical to support the interpolation from that since it wouldn't have license issues.

**Comment by philipl** [comment]
> Hi a bit off topic here, but do you have a ffmpeg binary to try out the filter with? Also, how does one specify the parameters in ffmpeg for the filter?

No, you'll need to build your own binary. As I said up top, the binaries cannot be distributed due to licence incompatibility - you *must* build your own binaries.

Parameter passing follows the normal documented ffmpeg filter syntax.

**Comment by philipl** [comment]
> AMD released the FSR3 source code, I wonder if it would be more practical to support the interpolation from that since it wouldn't have license issues and would support all vendors. EDIT: It looks like it only supports DX12 currently so it would need to be ported :/

Someone will probably poke around with it eventually but it's not clear to me if it can work without the detailed motion vectors that a game can produce (because it knows exactly where each object in the scene is going). For decoded video, we don't necessarily have any motion vectors, and even for codecs with such vectors, they are tracking where macroblocks are going and not scene elements.

**Comment by Jules-A** [comment]
> Someone will probably poke around with it eventually but it's not clear to me if it can work without the detailed motion vectors that a game can produce (because it knows exactly where each object in the scene is going). For decoded video, we don't necessarily have any motion vectors, and even for codecs with such vectors, they are tracking where macroblocks are going and not scene elements.

Yeah, I don't know about that either but I did notice this recently added to AMF: https://github.com/GPUOpen-LibrariesAndSDKs/AMF/blob/c48e50ad6c8723c006b2c145d8fa49ecc0651022/amf/doc/AMF_FRC_API.pdf

Not sure how similar it would be to FSR3 but it looks interesting.

**Comment by Jules-A** [comment]
> @philipl How can I make it do container_fps*2 Also do I have to enable display-resample, tscale and interpolation. I asked it since we pass it as vf

Pretty sure MPV will just see the video with twice the framerate of the original and none of those options are required. Although I am curious to how interpolation works when combined.

**Comment by philipl** [comment]
Right. This filter can do any arbitrary target frame rate, so you shouldn't need to use any other mechanisms. Eg: you can do 24->60 with it and, in theory, you will get proper 60fps frame pacing. But if you do decide to 24->48 plus interpolation, it would work fine too.

As for container fps substitution, I think that requires additional logic. I can look at it, but no promises.

**Comment by philipl** [comment]
Actually, I'm reopening this just to improve visibility for anyone who wants to use this.

**Comment by philipl** [comment]
I cannot provide a detailed guide on how to ensure you have all the required dependencies, but for this specific case, you must place `NvOFFRUC.h` in the `libavfilter/` directory next to the filter code.

You must have https://github.com/FFmpeg/nv-codec-headers  installed as well, if you haven't already (and maybe your distro has a package for it)

**Comment by Usulyre** [comment]
> I cannot provide a detailed guide on how to ensure you have all the required dependencies, but for this specific case, you must place `NvOFFRUC.h` in the `libavfilter/` directory next to the filter code.
> 
> You must have https://github.com/FFmpeg/nv-codec-headers installed as well, if you haven't already (and maybe your distro has a package for it)

I am trying to cross compile this using this script here:

https://github.com/rdp/ffmpeg-windows-build-helpers

Does NvOFFRUC.h have to be modified in some way? Or do I just place it in the libavfilter folder?

Do I need any other files to placed in any certain folders?

I just can't find any compiling instructions here for this.

I have installed the nv-codec-headers as well.

Anything else to add to make this compile correctly?

Thanks in advance.

**Comment by philipl** [comment]
Yes, you do need to modify it, as documented in the code: https://github.com/philipl/FFmpeg/blob/nvoffruc/libavfilter/vf_nvoffruc.c#L42

**Comment by Usulyre** [comment]
> Yes, you do need to modify it, as documented in the code: https://github.com/philipl/FFmpeg/blob/nvoffruc/libavfilter/vf_nvoffruc.c#L42

Ok I did follow that but still getting errors like below.

`In file included from libavfilter/vf_nvoffruc.c:48:
libavfilter/NvOFFRUC.h:215:5: error: expected '=', ',', ';', 'asm' or '__attribute__' before '{' token
  215 |     {
      |     ^
In file included from libavfilter/vf_nvoffruc.c:48:
libavfilter/NvOFFRUC.h:244:9: error: unknown type name 'NvOFFRUCResourceType'
  244 |         NvOFFRUCResourceType        eResourceType;      /**< [in]: Specifies whether resource created by client is DX or CUDA. */
      |         ^~~~~~~~~~~~~~~~~~~~
libavfilter/NvOFFRUC.h:245:9: error: unknown type name 'NvOFFRUCSurfaceFormat'
  245 |         NvOFFRUCSurfaceFormat       eSurfaceFormat;     /**< [in]: Specifies surface format i.e. NV12 or ARGB. */
      |         ^~~~~~~~~~~~~~~~~~~~~
libavfilter/NvOFFRUC.h:246:9: error: unknown type name 'NvOFFRUCCUDAResourceType'
  246 |         NvOFFRUCCUDAResourceType    eCUDAResourceType;  /**< [in]: Specifies whether CUDA resource is cuDevicePtr or cuArray. */
      |         ^~~~~~~~~~~~~~~~~~~~~~~~
libavfilter/vf_nvoffruc.c: In function 'process_work_frame':
libavfilter/vf_nvoffruc.c:291:58: warning: format '%lu' expects argument of type 'long unsigned int', but argument 4 has type 'int64_t' {aka 'long long int'} [-Wformat=]
  291 |             av_log(ctx, AV_LOG_DEBUG, "Matched f0: pts %lu\n", work_pts);
      |                                                        ~~^     ~~~~~~~~
      |                                                          |     |
      |                                                          |     int64_t {aka long long int}
      |                                                          long unsigned int
      |                                                        %llu
libavfilter/vf_nvoffruc.c:294:58: warning: format '%lu' expects argument of type 'long unsigned int', but argument 4 has type 'int64_t' {aka 'long long int'} [-Wformat=]
  294 |             av_log(ctx, AV_LOG_DEBUG, "Matched f1: pts %lu\n", work_pts);
      |                                                        ~~^     ~~~~~~~~
      |                                                          |     |
      |                                                          |     int64_t {aka long long int}
      |                                                          long unsigned int
      |                                                        %llu
libavfilter/vf_nvoffruc.c:297:57: warning: format '%lu' expects argument of type 'long unsigned int', but argument 4 has type 'int64_t' {aka 'long long int'} [-Wformat=]
  297 |             av_log(ctx, AV_LOG_DEBUG, "Unmatched pts: %lu\n", work_pts);
      |                                                       ~~^     ~~~~~~~~
      |                                                         |     |
      |                                                         |     int64_t {aka long long int}
      |                                                         long unsigned int
      |                                                       %llu
make: *** [ffbuild/common.mak:81: libavfilter/vf_nvoffruc.o] Error 1
`

**Comment by philipl** [comment]
That suggests you might have mangled something. Line 215 in the original flle in innocuous. make sure you didn't actually delete something else.

**Comment by Usulyre** [comment]
> That suggests you might have mangled something. Line 215 in the original flle in innocuous. make sure you didn't actually delete something else.

OK.

Well i got this error originally:

In file included from libavfilter/vf_nvoffruc.c:48:
libavfilter/NvOFFRUC.h:25:14: fatal error: Windows.h: No such file or directory
   25 |     #include <Windows.h>

However after  a bit of research I tried to rename a file "windows.h" to uppercase which leads to those errors i guess? It's in the cross compilers that the script I have uses.

Or is it something else?

**Comment by philipl** [comment]
I'm afraid I don't know enough about the windows build to help with that. Certainly you should modify NvOFFRUC.h rather than rename the file - which might cause problems elsewhere. But windows is supposed to be case insensitive, so it shouldn't make a difference.

**Comment by Usulyre** [comment]
> Ok. I've got a working [implementation](https://github.com/philipl/FFmpeg/tree/fruc-me) based on `vf_framerate` (which is a simpler blending filter). As noted, the licensing here is dubious - I do not believe any resulting binaries are distributable, and you need to provide libNvOFFRUC.so and libcuda-rt in your library path. But feel free to try it out.
> 
> ```
> mpv --vf=nvoffruc:fps=120 somefile.mkv
> ```

Do I need "libNvOFFRUC.so and libcuda-rt in my library path" for compilation?

**Comment by philipl** [comment]
No, but unfortunately, I am reminded of https://github.com/philipl/FFmpeg/blob/nvoffruc/libavfilter/vf_nvoffruc.c#L318

There is no logic for doing dynamic symbol loading on Windows.It's not particularly difficult as you can use `compat/w32dlfcn.h` to probably make the existing `dlopen` and `dlsym` work as-is. But you'll need to do it. Sorry.

**Comment by Usulyre** [comment]
> No, but unfortunately, I am reminded of https://github.com/philipl/FFmpeg/blob/nvoffruc/libavfilter/vf_nvoffruc.c#L318
> 
> There is no logic for doing dynamic symbol loading on Windows.It's not particularly difficult as you can use `compat/w32dlfcn.h` to probably make the existing `dlopen` and `dlsym` work as-is. But you'll need to do it. Sorry.

Not sure once I get around those strange errors, that I understand what to do for dynamic symbol loading.

**Comment by SephirothFFKH** [comment]
I think another approach was the vapoursynth route it which reminds me of a old project with the legacy NV Optical Flow [vapoursynth-nvof](https://bitbucket.org/mystery_keeper/vapoursynth-nvof/src/master/) but the maintainer @mystery-keeper never upgraded that project to NV OFFRUC unfortunately.

**Comment by philipl** [comment]
> I think another approach was the vapoursynth route which reminds me of an old project with the legacy NV Optical Flow [vapoursynth-nvof](https://bitbucket.org/mystery_keeper/vapoursynth-nvof/src/master/) but the maintainer @mystery-keeper never upgraded that project to NV OFFRUC unfortunately.

The vapoursynth integration doesn't support passing hardware frames so the overhead of copying back and forth between system and GPU memory multiple times rather undermines the whole approach.

**Comment by Usulyre** [comment]
> No, but unfortunately, I am reminded of https://github.com/philipl/FFmpeg/blob/nvoffruc/libavfilter/vf_nvoffruc.c#L318
> 
> There is no logic for doing dynamic symbol loading on Windows.It's not particularly difficult as you can use `compat/w32dlfcn.h` to probably make the existing `dlopen` and `dlsym` work as-is. But you'll need to do it. Sorry.

I'm still having issue with the NvOFFRUC.h file can you help out with that?

**Comment by Usulyre** [comment]
> > No, but unfortunately, I am reminded of https://github.com/philipl/FFmpeg/blob/nvoffruc/libavfilter/vf_nvoffruc.c#L318
> > There is no logic for doing dynamic symbol loading on Windows.It's not particularly difficult as you can use `compat/w32dlfcn.h` to probably make the existing `dlopen` and `dlsym` work as-is. But you'll need to do it. Sorry.
> 
> I'm still having issue with the NvOFFRUC.h file can you help out with that?

Anybody manage to get this to work in windows os?

**Comment by philipl** [comment]
You can ask the author of the patch which added windows build support.

**Comment by Usulyre** [comment]
> You can ask the author of the patch which added windows build support.

Hi,

I did ask him already, but the issue is I am having is with NvOFFRUC.h file, just trying to follow what you mentioned to edit, the errors  I got above still are there, I really don't know how to get pass the errors.

**Comment by Usulyre** [comment]
> You can ask the author of the patch which added windows build support.

Ok, I have managed to compile shared ffmpeg and mpv with nvoffruc for windows os and in the directory with ffmpeg and mpv binaries I put in the same folder with these two files:  cudart64_110.dll and NvOFFRUC.dll .

However, I am getting an error like this:

"[ffmpeg] Impossible to convert between the formats supported by the filter 'mpv_src_default_in' and the filter 'auto_scale_0'
[lavfi] failed to configure the filter graph"

Maybe I don't have something correct in those two files that I modified: NvOFFRUC.h and vf_nvoffruc.c ?

Anybody care to look at them? I can upload the modified files here if that is ok?

**Comment by philipl** [comment]
That just means you didn't configure mpv properly at runtime.

At a minimum you need `mpv --hwdec=nvdec --vf=nvoffruc=fps=<some value>`

**Comment by Usulyre** [comment]
> That just means you didn't configure mpv properly at runtime.
> 
> At a minimum you need `mpv --hwdec=nvdec --vf=nvoffruc=fps=<some value>`

I sitll got this error:

![Image](https://github.com/user-attachments/assets/e7150834-569c-4ed1-980e-dcc94a6b2c17)

**Comment by philipl** [comment]
Yeah, I don't think mpv will try and hardware decode MPEG-4 ASP by default. That's not a great example file to use. You should use something less ancient (H.264, HEVC, VP9. AV-1, etc). You can read the docs on that.

If you really want to test with a software decoded format, you'd need to use

`-vf=format=cuda,nvoffruc=fps=120` to tell it to upload the decoded frames to the GPU

**Comment by Usulyre** [comment]
> Yeah, I don't think mpv will try and hardware decode MPEG-4 ASP by default. That's not a great example file to use. You should use something less ancient (H.264, HEVC, VP9. AV-1, etc). You can read the docs on that.
> 
> If you really want to test with a software decoded format, you'd need to use
> 
> `-vf=format=cuda,nvoffruc=fps=120` to tell it to upload the decoded frames to the GPU

I am still working on getting mpv compiled, but I got static build of ffplay and ffmpeg built with nvoffruc I had to enable vulkan and libplacebo for the build.

Is static build ok or is shared better for this filter?

So far, this cmd works, i tried 120 and 144 fps but the playback wasn't at all smooth.

ffplay -hwaccel cuda -vf "nvoffruc=fps=60" 1.avi

![Image](https://github.com/user-attachments/assets/f70427da-a6bd-43a2-9ad4-0eeac207cd4c)
![Image](https://github.com/user-attachments/assets/b9a2b389-01e7-4e4f-bb45-749cd0a1fc59)
![Image](https://github.com/user-attachments/assets/47b1bbfa-290f-4e9e-930a-653c9a43696c)
![Image](https://github.com/user-attachments/assets/fec30cbd-8edc-4bde-8279-f4ca340da72b)

**Comment by philipl** [comment]
🤷 Not sure what to say - if the filter is running, then after that it's nvidia's business as to what actually happens.

But I will say you are not testing with the simplest possible samples and command lines. You should start with a 24fps h.264 sample and do:

`mpv --hwdec=nvdec --vf=nvoffruc=fps=48 sample.mkv`

It will be really obvious if it's working and you can frame step through it and see the interpolated frames - it's obvious.

**Comment by philipl** [comment]
> from https://github.com/NVIDIA-Omniverse/PhysX/discussions/384
> 
> In addition, we’re also open-sourcing the full GPU compute shader implementation of the Flow SDK, our real-time, sparse grid–based fluid simulation library.
> 
> Does that help in any way?

No. It's an unrelated library.

**Comment by Usulyre** [comment]
> 🤷 Not sure what to say - if the filter is running, then after that it's nvidia's business as to what actually happens.
> 
> But I will say you are not testing with the simplest possible samples and command lines. You should start with a 24fps h.264 sample and do:
> 
> `mpv --hwdec=nvdec --vf=nvoffruc=fps=48 sample.mkv`
> 
> It will be really obvious if it's working and you can frame step through it and see the interpolated frames - it's obvious.

Ok, so I cross compiled mpv now and got nvoffruc working and with different video files that don't work because of the pixel format issue I was having.

Examples:

mpv --hwdec=nvdec --gpu-api=opengl --vf=format=cuda,nvoffruc=fps=120 NV12.mp4

![Image](https://github.com/user-attachments/assets/b09f889e-8547-44de-8a7f-c40313bfacc2)



mpv --hwdec=nvdec --gpu-api=opengl --framedrop=no --vf=lavfi-format=pix_fmts=nv12,format=cuda,nvoffruc=fps=60 1.avi

![Image](https://github.com/user-attachments/assets/67bed287-d84e-4a84-815b-d722e4e76a94)

Capture card example:

mpv av://dshow:video="AVerMedia ExtremeCap UVC" --demuxer-lavf-o=pixel_format=yuyv422,rtbufsize=702000k,video_size=720x480,framerate=60 --framedrop=no --hwdec=nvdec --gpu-api=opengl --geometry=1280x720 --container-fps-override=24 --vf=lavfi-format=pix_fmts=nv12,fps=fps=24,format=cuda,nvoffruc=fps=60



![Image](https://github.com/user-attachments/assets/76895340-3b9c-4fec-a8b4-3d4a20c46914)

**Comment by Usulyre** [comment]
> 🤷 Not sure what to say - if the filter is running, then after that it's nvidia's business as to what actually happens.
> 
> But I will say you are not testing with the simplest possible samples and command lines. You should start with a 24fps h.264 sample and do:
> 
> `mpv --hwdec=nvdec --vf=nvoffruc=fps=48 sample.mkv`
> 
> It will be really obvious if it's working and you can frame step through it and see the interpolated frames - it's obvious.

Is it ok to post/upload the vf_nvoffruc.c and NvOFFRUC.h as text files here? I had a really hard time to get it to work, and I would like to contribute the changes to whomever wants to use this filter in ffmpeg/mpv for windows os. 

I would also like to point out that the that your latest vf_nvoffruc.c does not include the changes for the windows support, I had to manually input everything - copy/paste.

Also, the NvOFFRUC.h file had additional changes made from what you mentioned as I kept getting errors when compiling.

**Comment by philipl** [comment]
Please send a PR to my GitHub fork with the changes. I thought I got all the windows changes in but maybe I missed something.

**Comment by mczhao1977** [comment]
> > Someone will probably poke around with it eventually but it's not clear to me if it can work without the detailed motion vectors that a game can produce (because it knows exactly where each object in the scene is going). For decoded video, we don't necessarily have any motion vectors, and even for codecs with such vectors, they are tracking where macroblocks are going and not scene elements.
> 
> Yeah, I don't know about that either but I did notice this recently added to AMF: https://github.com/GPUOpen-LibrariesAndSDKs/AMF/blob/c48e50ad6c8723c006b2c145d8fa49ecc0651022/amf/doc/AMF_FRC_API.pdf
> 
> Not sure how similar it would be to FSR3 but it looks interesting.

So what's the current status about FSR3 or AMF integration? FSR3 has a sprarate optical flow implementation which can provide MV.

**Comment by Usulyre** [comment]
> Please send a PR to my GitHub fork with the changes. I thought I got all the windows changes in but maybe I missed something.

Can't open a new pull request in your ffmpeg fork, only for collaborators?

**Comment by philipl** [comment]
> Can't open a new pull request in your ffmpeg fork, only for collaborators?

I don't know what you're seeing, but it's not restricted. Anyone should be able to open a PR

**Comment by philipl** [comment]
Weird. It seems to have inherited some restriction from the original ffmpeg repo that I cannot alter. Well, use `git format-patch` to get a patch I can apply and attach it here (or upload it as a gist, etc) and I'll add it.

**Comment by Usulyre** [comment]
![Image](https://github.com/user-attachments/assets/28669581-2121-477e-b56f-82d167a59ca6)

**Comment by Usulyre** [comment]
> Weird. It seems to have inherited some restriction from the original ffmpeg repo that I cannot alter. Well, use `git format-patch` to get a patch I can apply and attach it here (or upload it as a gist, etc) and I'll add it.

https://gist.github.com/Usulyre/e8c2defefeb0ec7245beba8d76aaef44

**Comment by philipl** [comment]
> https://gist.github.com/Usulyre/e8c2defefeb0ec7245beba8d76aaef44

Thanks, but why did you make the error string array `static`?

**Comment by Usulyre** [comment]
> > https://gist.github.com/Usulyre/e8c2defefeb0ec7245beba8d76aaef44
> 
> Thanks, but why did you make the error string array `static`?

It was changed by copilot-github, I couldn't get past the errors that kept coming up without the changes.

**Comment by philipl** [comment]
Can you show me the error you get without that specific change?

**Comment by Usulyre** [comment]
> Can you show me the error you get without that specific change?

Which file are you referring to by the way?

**Comment by philipl** [comment]
> > Can you show me the error you get without that specific change?
> 
> Which file are you referring to by the way?

In NvOFFRUC.h, you added a `static` to `static const char *NvOFFRUCErrorString[] = {`

That should not be required.

**Comment by Usulyre** [comment]
> > > Can you show me the error you get without that specific change?
> > 
> > 
> > Which file are you referring to by the way?
> 
> In NvOFFRUC.h, you added a `static` to `static const char *NvOFFRUCErrorString[] = {`
> 
> That should not be required.

Got the error below:

![Image](https://github.com/user-attachments/assets/fcbe7ca5-6c62-4ea3-b8b2-a01b97396697)

**Comment by philipl** [comment]
That seems weird, but remove the `static` from the working file and try again. Leave the other changes in place.

**Comment by Usulyre** [comment]
> That seems weird, but remove the `static` from the working file and try again. Leave the other changes in place.

Do I remove this:       = {     

and then recompile?

If so, the screenshot was taken when I did that an hour ago.

**Comment by philipl** [comment]
No, do not remove that. Just remove the `static`.

**Comment by Usulyre** [comment]
> No, do not remove that. Just remove the `static`.

Ok I will try that, but I didn't add 'static' it is part of the original file, right?

**Comment by philipl** [comment]
> > No, do not remove that. Just remove the `static`.
> 
> Ok I will try that, but I didn't add 'static' it is part of the original file, right?

The `static` is not in my copy of the file.

**Comment by philipl** [comment]
I've updated my tree and added the link to it to my first reply to this issue. It's basically impossible for anyone to find the link in the reply chain at this point. Thank you @Usulyre!

**Comment by Usulyre** [comment]
> I've updated my tree and added the link to it to my first reply to this issue. It's basically impossible for anyone to find the link in the reply chain at this point. Thank you [@Usulyre](https://github.com/Usulyre)!

Would you be able to add the changes into vf_nvoffruc.c that I put in my gist file? I noticed that the windows support is still not present in your fork.

**Comment by philipl** [comment]
> Would you be able to add the changes into vf_nvoffruc.c that I put in my gist file? I noticed that the windows support is still not present in your fork.

Whoops. Done.

**Comment by FrozenGalaxy** [comment]
Can someone summarize the current status and current implementation of this?
I'd love to try it out (on windows).

Slightly offtopic, beside this and mvtools, is there any other modern motion interpolation solution available for mpv?

**Comment by Juraj-Masiar** [comment]
@FrozenGalaxy  I was using SVP for years but now I'm using "Lossless Scaling".
It's not fully automatic but you can define a shortcut like "F1" to activate it in currently focused app (MPV).
The result is quite good - much better than CPU interpolation in SVP, maybe a bit worse than RIFE interpolation (especially around subtitles), but it consumes fraction of GPU and it supports even AMD GPUs. I can do 6x24fps (144fps) at 4K, which is unreal with RIFE.
Plus it works in any other app, even YouTube player.

**Comment by FrozenGalaxy** [comment]
> I'm using "Lossless Scaling".
Sadly its not open source and paywalled?

**Comment by philipl** [comment]
> Can someone summarize the current status and current implementation of this? I'd love to try it out (on windows).

Current status is it works with the limitations discussed in the early comments. If you can build ffmpeg, you should be able to build this branch - but you'll have to build it yourself.

**Comment by Usulyre** [comment]
> > Can someone summarize the current status and current implementation of this? I'd love to try it out (on windows).
> 
> Current status is it works with the limitations discussed in the early comments. If you can build ffmpeg, you should be able to build this branch - but you'll have to build it yourself.

Just wondering, 

![Image](https://github.com/user-attachments/assets/6647b40c-8fbc-44e7-8d54-b53a8f794fe1)

Is there a way to change the optical flow grid size?

**Comment by philipl** [comment]
> > > Can someone summarize the current status and current implementation of this? I'd love to try it out (on windows).
> > 
> > 
> > Current status is it works with the limitations discussed in the early comments. If you can build ffmpeg, you should be able to build this branch - but you'll have to build it yourself.
> 
> Just wondering,
> 
> ![Image](https://github.com/user-attachments/assets/6647b40c-8fbc-44e7-8d54-b53a8f794fe1)
> 
> Is there a way to change the optical flow grid size?

Yes, you can read the SDK docs on how to do it and adjust the code. But I never saw any meaningful difference when I tried changing it.

**Comment by FrozenGalaxy** [comment]
I managed to get this working on **latest FFmpeg (git master) + Windows**, but it was honestly quite messy to set up.

---

## Requirements

You need to obtain NVIDIA’s Optical Flow SDK yourself:

- Header:  
  https://developer.nvidia.com/opticalflow/download  
- Runtime:  
  `NvOFFRUC.dll`

---

## FFmpeg filter patch

Philip’s original filter does **not compile or work with current FFmpeg**, so it needs patching.

I adapted it to current FFmpeg APIs (mainly hwcontext / filter graph changes).

**Diff against current `ffmpeg.git master`:**  
https://gist.github.com/FrozenGalaxy/a35af56c1ca97de7de6079fb7347bdc2

> **Note:**  
> This was largely AI-assisted and I’m not very experienced with C/C++, so it’s definitely not pretty - but it works.

EDIT: I optimized it with AI a bit to get more performance, but it's still not good enough, on a 4090, it struggles with 4kHDR sources (to 60FPS) (silly AI optimize version: https://gist.github.com/FrozenGalaxy/51d726aff9213d986e296958faa2a805)

---

## NVIDIA header patch

The NVIDIA header (`NvOFFRUC.h`) also needs modification to compile in a pure C environment:

- remove `using namespace std;`
- replace `bool*` with `void*`
- minor compatibility fixes

I’m **not distributing the header itself**, only a patch:

https://gist.github.com/FrozenGalaxy/fa62bfbd24d511b512c6826b5932ff01

> I *think* this should be fine licensing-wise since it’s just a diff - but I’m not 100% sure.

---

## Running it (mpv example)

This finally worked for me:

```bash
mpv.com --vf=scale=w=iw:h=ih:flags=fast_bilinear,format=nv12,hwupload_cuda,nvoffruc=fps=60 video.mkv
```

Note: Even when it works, the overall quality isn’t great - especially in slower scenes, motion can look unstable or “jittery” (e.g. heads subtly wobbling), which makes it feel noticeably less smooth than expected._

My verdict: So far nothing beats SVP Pro (for 4k+) and mvtools for 1080p media

_(god, I also AI formatted this text, it feels so bad lol, but hey ho still easier to read than my mess before)_

**Comment by philipl** [comment]
If you want to send me an actual diff against the branch in my github fork, you can. I've force-pushed it back to a roughly up-todate ffmpeg master. I don't know why you needed to apparently make a lot of changes to get it to compile on windows, given the whole historical saga of getting it to work that's up thread.

In terms of performance, I don't know what the maximum realistic throughput is, but the filter is very deliberately designed for flexibility rather than performance - it reinitialises on every frame to keep a stateless model rather than trying to re-use input frames. This allows for arbitrary output framerates vs fixed 2x or whatever (like Lossless Scaling FG vs DLSS FG) but it incurs more overhead.

Perhaps most importantly, it's unclear if this works on Blackwell GPUs. nvidia have made no official statement and I can't get it to work on my card. That means I currently have no way to test anything.

If anyone here has access to both 50 series and older, but supported, hardware, it would be great if you could confirm what works and what doesn't.

**Comment by philipl** [comment]
Good news and bad news on the 50 series support!

The issue is that Blackwell GPUs cannot run the compiled cuda kernels for older hardware. The binaries that nvidia are distributing haven't been updated for Blackwell, but the source code actually works (I was given access to the library source, but obviously any fixes I come up with don't help anyone else). I was able to recompile for blackwell and it runs. So at least there isn't a hardware incompatibility issue.

**Comment by FrozenGalaxy** [comment]
In any case the performance is subpar and that's all I wanted to check actually, sadly. 
I tested it on a 4k hdr source with target 60fps as mentioned above, where it struggles on some scenes on my 4090.

**Comment by Usulyre** [comment]
> In any case the performance is subpar and that's all I wanted to check actually, sadly. I tested it on a 4k hdr source with target 60fps as mentioned above, where it struggles on some scenes on my 4090.

If anyone is interested, there's 

 https://github.com/HopperLogger/mpv-frame-interpolator

I managed to compile a windows version as well.

https://github.com/HopperLogger/HopperRender/issues/20

**Comment by philipl** [comment]
Why would someone fork mpv to create a filter? Just send a PR, or contribute it to ffmpeg which is where filter implementations should go anyway.

**Comment by Brainiarc7** [comment]
FFmpeg's libplacebo filter has the [same feature-set](https://superuser.com/a/1819380/473795), and it's performant enough. So yes, forking mpv to re-create a filter for this is unnecessary. [Custom shaders](https://libplacebo.org/glsl/) can [also be loaded for libplacebo on runtime](https://stackoverflow.com/a/76793322) for further refinements. The software-based [minterpolate](https://ffmpeg.org/ffmpeg-filters.html#toc-minterpolate) filter is the gold standard for this, though.

**Comment by Usulyre** [comment]
> FFmpeg's libplacebo filter has the [same feature-set](https://superuser.com/a/1819380/473795), and it's performant enough. So yes, forking mpv to re-create a filter for this is unnecessary. [Custom shaders](https://libplacebo.org/glsl/) can [also be loaded for libplacebo on runtime](https://stackoverflow.com/a/76793322) for further refinements. The software-based [minterpolate](https://ffmpeg.org/ffmpeg-filters.html#toc-minterpolate) filter is the gold standard for this, though.

I think that ffmpeg libplacebo filter works the same as the --interpolation in mpv ? I've tried both, and I don't see any difference.

Meaning frame interpolation (soap opera effect) like what nvoffruc or mpv frame interpolator or svp or LSFG, amd fluid motion or afmf or intel vpl ai frame interpolation etc do.

**Comment by lauyv** [comment]
https://github.com/mpv-player/mpv/pull/18426

wait this be merged...

**Comment by philipl** [comment]
The ffmpeg change that's merged is a pure vulkan implementation where I had to provide my own interpolation algorithm. So there's no NVIDIA code with a bad licence. there's just that small additional change required for it to work in mpv.

**Comment by philipl** [comment]
With #18426 merged, I'm going to close this one. We never merged, and never will merge, support for the nvoffruc library due to the licensing, but the Vulkan filter uses the same underlying optical flow hardware, in combination with an independently developed interpolation algorithm that is dumber, but doesn't result in massively worse quality. That algorithm can be improved over time by motivated people.

**Comment by lauyv** [comment]
Thanks! It’s working now. I’m using the following configuration:
`Ctrl+1 vf toggle format=fmt=vulkan,fruc_vulkan=fps=60=auto=slow; show-text "toggle FRUC"`

**Comment by philipl** [comment]
You would want to use `fruc_vulkan=fps=60:perf=slow` but `perf` defaults to `slow` anyway.

**Comment by lauyv** [comment]
> You would want to use `fruc_vulkan=fps=60:perf=slow` but `perf` defaults to `slow` anyway.您应该使用 fruc_vulkan=fps=60:perf=slow，但 perf 默认值为 slow。

okay, thanks for reminding me. I'll test it when I get off work.

edit:
```ini
[fruc-vulkan]
profile-cond=platform == "windows" and container_fps > 0 and container_fps < 60 and height <= 1080
profile-restore=copy
vf-append=@frc-input:format=fmt=vulkan
vf-append=@frc:fruc_vulkan=fps=[source_fps*3]:perf=slow
```

It works well overall, but I occasionally get severe visual artifacts during scene cuts.

Does `fruc_vulkan` have any scene-change detection, or does it rely entirely on `VK_NV_optical_flow` to handle cuts?
